### PR TITLE
Support string concatenation.

### PR DESCRIPTION
**Pull request by Moderocky**
## Summary

1. Allows the joining of **strings** with `+`, e.g. `"foo" + "bar" = "foobar"`. \
Does **not** allow the joining of strings and numbers or other non-coercible objects.
2. Adds a var-args `concat(...)` function that permits joining anything into a single string (including non-texts).

### Description

I'm not a fan of strings as they are in Skript: I dislike that there's no true literal, I dislike how cumbersome and inefficient `"%...%"` interpolation is (both to write and to actually evaluate).
Aside from interpolation, our only other option for joining strings is the expression `(concat[enate]|join) %texts% [(with|using|by) [[the] delimiter] %text%]` which has a rather unpleasant pattern (it sounds more like an effect than an expression).
This is designed for joining texts, but _it can't coerce types properly_, so you still end up needing to use interpolation (or store things in middle man variables) to actually join things that aren't strings (e.g. `"your lucky number is "` and `66`).

### What This Doesn't Do

In 2.7 or 2.8 (I forget which) a deliberate change was made to make illegal expressions in maths completely fail (evaluate to `<none>`) rather than be permitted as `0`.
**This doesn't change that**. This doesn't permit you to add `"hello"` and `81`: it will fail at parsing time and, even if you try and get around that with variables, it will still evaluate to `<none>` as intended.

String addition ONLY supports adding strings together, whether they're from expressions, variables or `"..."`.

```applescript
broadcast "hello" + "world" # OK!

set {var} to "..."
broadcast "hello" + {var} # OK

broadcast "hello" + [some string expression] # OK
```

```applescript
broadcast "hello" + 1 # NOT OK! Fails at parse time

set {var} to 1
broadcast "hello" + {var} # NOT OK! Bad math to <none>

broadcast "hello" + [some non-string expression] # NOT OK! Probably fails at parse time
```

### String Addition

This pull request adds string addition arithmetic: `"A" + "B" = "AB"`.

Lots of languages allow you to concatenate (glue together) strings. Most languages use `+`, some like to do their own thing with `.` or `&` or some other joiner.
Just because other languages do something isn't a sufficient reason to add it to Skript, though.

I like it as a feature:
 - I think it's arguably cleaner and neater than `"%...%"` (especially if you have multiple interpolations in one text!)
 - It's especially cleaner than trying to use an expression (and heaven forbid a _text_) inside an interpolation, where you're having to double your quotes and whatever.
 - I think, realistically, it's more immediately understandable to a beginner that `"a" + "b" = "ab"` than whatever's going on with `"a%""b""%"`.
 - It _ought_ to be safer/easier for the machine to interpret, interpolation's usually pretty heavy (although this is Skript, who knows).

### String `concat(...)` Function

This pull request adds a `concat(...)` variable arguments function.
This is more advanced than the `concatenate|join` expression, since it properly supports non-text parameters even without coercion.

Concat takes in multiple objects and returns the result of all of them stringified and glued together.

```applescript
concat("hello") = "hello"
concat("hello", " world") # "hello world"
concat("hello", " ", "there") # "hello there"

set {_var} to "hello"
concat({_var}, " world") # "hello world"
```

Concat allows the joining of strings and objects, which are run through `Classes.toString`.

```applescript
concat("foo", 1) # "foo1"
concat("my nice new ", stone sword) # "my nice new stone sword"
```

Concat **ALWAYS** returns a text, even if none of the inputs are text. It doesn't do maths evaluation, it purely joins things into a text. 

```applescript
concat(1, 2) # "12"
concat(1, 2, 3) # "123"
concat(stone sword, stone sword) # "stone swordstone sword"
```

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

**Review by AyhamAl-Ali — APPROVED** [review]
Wonderful 🎉 Let's hope it doesn't break anything we didn't think of

File: src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java [hunk 1]
```diff
@@ -568,6 +569,24 @@ public Boolean[] executeSimple(Object[][] params) {
 		}).description("Returns true if the input is NaN (not a number).")
 			.examples("isNaN(0) # false", "isNaN(0/0) # true", "isNaN(sqrt(-1)) # true")
 			.since("2.8.0");
+
+		Functions.registerFunction(new SimpleJavaFunction<String>("concat", new Parameter[]{
+			 new Parameter<>("texts", DefaultClasses.OBJECT, false, null)
+		},DefaultClasses.STRING, true) {
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/classes/data/DefaultFunctions.java:575]:
```suggestion
		Functions.registerFunction(new SimpleJavaFunction<String>("concat", new Parameter[] {
			 new Parameter<>("texts", DefaultClasses.OBJECT, false, null)
		}, DefaultClasses.STRING, true) {
```

File: src/main/java/ch/njol/skript/classes/data/DefaultOperations.java [hunk 2]
```diff
@@ -115,6 +115,9 @@ public class DefaultOperations {
 		Arithmetics.registerOperation(Operator.SUBTRACTION, Date.class, Timespan.class, Date::minus);
 		Arithmetics.registerDifference(Date.class, Timespan.class, Date::difference);
 
+		// String + String
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/classes/data/DefaultOperations.java:118]:
```suggestion
		// String - String
```
The comment describes the type of the two operands, not the operation

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]
For `string + string` should we treat an empty variable as an empty string? E.x. `"some string" + {_none} = "some string"`

File: src/test/skript/tests/misc/string concatenation.sk [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	# variable-string + string
+	set {_var} to "hello"
+	assert "%{_var}% " + "there" is "hello there" with "var-string + string concat failed"
+
+	# string + number = <none>
+	# parser prevents us adding "foo" + 1 or comparing "foo" + 1 with a string
+	# we test the edge case where somebody slips past us!
+	set {_var} to 1
+	set {_var} to "foo" + {_var}
+	assert {_var} doesn't exist with "string + number concat succeeded? %{_var}%"
```
Comment (sovdeeth) [inline src/test/skript/tests/misc/string concatenation.sk:29]: you could use the new expected/got suppliers!

File: src/test/skript/tests/syntaxes/functions/concat.sk [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+	# variable-string + string
+	set {_var} to "hello"
+	assert concat("%{_var}% ", "there") is "hello there" with "var-string + string concat() failed"
+
+	# string + non-string
+	# unlike the maths expression we CAN concat objects here!
+	set {_var} to 1
+	set {_var} to concat("foo", {_var})
+	assert {_var} is "foo1" with "string + number concat() failed: %{_var}%"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/functions/concat.sk:29]: could remove {_var} from error message, it will be supplied automatically

**Review by sovdeeth — APPROVED** [review]

File: src/test/skript/tests/syntaxes/functions/concat.sk (hunk 4 above)
Comment (Moderocky) [reply src/test/skript/tests/syntaxes/functions/concat.sk:29]: Only since today 😔 I shall do so

**Comment by UnderscoreTud** [comment]
> > For `string + string` should we treat an empty variable as an empty string? E.x. `"some string" + {_none} = "some string"`
> 
> Do you have a suggestion? I feel like it shouldn't return nothing as the result like with invalid math.

You can do that by registering a default value for strings using `Arithmetics.registerDefaultValue)`

**Review by UnderscoreTud — APPROVED** [review]

**Comment by Moderocky** [comment]
> You can do that by registering a default value for strings using `Arithmetics.registerDefaultValue`

Yes I know, I was asking what you think the default should be (if not what it is now).

**Comment by UnderscoreTud** [comment]
> > You can do that by registering a default value for strings using `Arithmetics.registerDefaultValue`
> 
> Yes I know, I was asking what you think the default should be (if not what it is now).

You can just make it an empty string

**Pull request merged**